cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -0,0 +1,179 @@
+
+  @Inject
+  ArtifactRepository(CConfiguration cConf, ArtifactStore artifactStore) {
+    this.tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    this.artifactStore = artifactStore;
+    this.artifactInspector = new ArtifactInspector(cConf);
+  }
+
+  @VisibleForTesting
+  void clear(Id.Namespace namespace) throws IOException {
+    artifactStore.clear(namespace);
+  }
+
+  /**
+   * Returns a {@link SortedMap} of all plugins available for the given artifact. The keys
+   * are sorted by the {@link ArtifactDescriptor} for the artifact that contains plugins available to the given
+   * artifact. Only unique {@link PluginClass} are returned in the value
+   * Collection, where uniqueness is determined by the plugin class type and name only.
+   *
+   * @param artifactId the id of the artifact to get plugins for
+   * @return an unmodifiable sorted map from plugin artifact to plugins in that artifact
+   * @throws IOException if there was an exception reading plugin metadata from the artifact store
+   */
+  public SortedMap<ArtifactDescriptor, List<PluginClass>> getPlugins(Id.Artifact artifactId) throws IOException {
+    return artifactStore.getPluginClasses(artifactId);
+  }
+
+  /**
+   * Returns a {@link Map.Entry} representing the plugin information for the plugin being requested.
+   *
+   * @param artifactId the id of the artifact to get plugins for
+   * @param pluginType plugin type name
+   * @param pluginName plugin name
+   * @param selector for selecting which plugin to use
+   * @return the entry found or {@code null} if none was found
+   * @throws IOException if there was an exception reading plugin metadata from the artifact store
+   * @throws PluginNotExistsException if no plugins of the given type and name are available to the given artifact
+   */
+  public Map.Entry<ArtifactDescriptor, PluginClass> findPlugin(Id.Artifact artifactId, final String pluginType,
+                                                               final String pluginName, PluginSelector selector)
+    throws IOException, PluginNotExistsException {
+    return selector.select(artifactStore.getPluginClasses(artifactId, pluginType, pluginName));
+  }
+
+  /**
+   * Inspects and builds plugin information for the given artifact.
+   *
+   * @param artifactId the id of the artifact to inspect and store
+   * @param artifactFile the artifact to inspect and store
+   * @param parentArtifacts artifacts the given artifact extends.
+   *                        If null, the given artifact does not extend another artifact
+   * @throws IOException if there was an exception reading from the artifact store
+   * @throws ArtifactRangeNotFoundException if none of the parent artifacts could be found
+   */
+  @VisibleForTesting
+  void inspectArtifact(Id.Artifact artifactId, File artifactFile, @Nullable Set<ArtifactRange> parentArtifacts)
```

Comment:
As discussed, this should be public and we can rename this method to addArtifact?